dashboard/test/lib/pd/survey_pipeline/daily_survey_retriever_test.rb

```diff
@@ -0,0 +1,111 @@
+      retriever = Pd::SurveyPipeline::DailySurveyRetriever.new
+
+      result = retriever.retrieve_data
+
+      assert_equal @workshop_submissions.length, result[:workshop_submissions]&.length
+      assert_equal @facilitator_submissions.length, result[:facilitator_submissions]&.length
+      assert_equal @survey_questions.length, result[:survey_questions]&.length
+    end
+
+    test 'can retrieve data using workshop id filter' do
+      filter = {workshop_ids: @workshops.first.id}
+      retriever = Pd::SurveyPipeline::DailySurveyRetriever.new filter
+
+      result = retriever.retrieve_data
+
+      assert_equal @workshop_submissions.length / @workshops.length,
+        result[:workshop_submissions]&.length
+      assert_equal @facilitator_submissions.length / @workshops.length,
+        result[:facilitator_submissions]&.length
+      assert_equal @workshop_form_ids.length + @facilitator_form_ids.length,
+        result[:survey_questions]&.length
+    end
+
+    test 'can retrieve data using form id filter' do
+      filter = {form_ids: @workshop_form_ids.first}
+      retriever = Pd::SurveyPipeline::DailySurveyRetriever.new filter
+
+      result = retriever.retrieve_data
+
+      assert_equal @workshop_submissions.length / @workshop_form_ids.length,
+        result[:workshop_submissions]&.length
+      assert_equal 0, result[:facilitator_submissions]&.length
+      assert_equal 1, result[:survey_questions]&.length
+    end
+
+    test 'can retrieve data using both workshop id and form id filters' do
+      filter = {workshop_ids: @workshops.first.id, form_ids: @facilitator_form_ids.first}
+      retriever = Pd::SurveyPipeline::DailySurveyRetriever.new filter
+
+      result = retriever.retrieve_data
+
+      assert_equal 0, result[:workshop_submissions]&.length
+      assert_equal @facilitator_submissions.size / (@workshops.length * @facilitator_form_ids.size),
+        result[:facilitator_submissions]&.length
+      assert_equal 1, result[:survey_questions]&.length
+    end
+
+    test 'return empty if workshop does not have submission' do
+      # Use non-existence workshop id as filter
+      filter = {workshop_ids: @workshops.pluck(:id).max + 1}
+      retriever = Pd::SurveyPipeline::DailySurveyRetriever.new filter
+
+      result = retriever.retrieve_data
+
+      assert_equal 0, result[:workshop_submissions]&.length
+      assert_equal 0, result[:facilitator_submissions]&.length
+      assert_equal 0, result[:survey_questions]&.length
+    end
+
+    test 'return empty if form does not have submission' do
+      # Use non-existence form id as filter
+      filter = {form_ids: @workshop_form_ids.max + 1}
+      retriever = Pd::SurveyPipeline::DailySurveyRetriever.new filter
+
+      result = retriever.retrieve_data
+
+      assert_equal 0, result[:workshop_submissions]&.length
+      assert_equal 0, result[:facilitator_submissions]&.length
+      assert_equal 0, result[:survey_questions]&.length
```

Comment:
Your tests are checking length of results, but I'd like them to also check the _shape_ of the results, since it's such an important part of how the whole pipeline fits together.  Can you take one of your tests that does return results and checks that each of the returned collections contains the expected model types?

Reply:
Resolve this for now as the data types will be abstracted away with the next Result Object update. Will add if we thing it's necessary after that change.